src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs

```diff
@@ -288,6 +288,41 @@ internal TelemetryProcessorChain TelemetryProcessorChain
             }
         }
 
+        /// <summary>
+        /// This method will set the connection string, set the instrumentation key, validate the endpoints, and set the TelemetryChannel.Endpoint.
+        /// </summary>
+        /// <param name="connectionString">Users unique connection string. This must include the Instrumentation Key. Format: "key1=value1;key2=value2;key3=value3".</param>
+        /// <exception cref="ConnectionStringDuplicateKeyException">Thrown when the connection string has duplicate keys.</exception>
+        /// <exception cref="ConnectionStringInvalidDelimiterException">Thrown when the connection string is formatted incorrectly.</exception>
+        /// <exception cref="ConnectionStringInvalidEndpointException">Thrown when the endpoint that was provided is invalid. Exception contains more information.</exception>
+        /// <exception cref="ConnectionStringMissingInstrumentationKeyException">Thrown when the connection string does not contain an instrumentation key (required).</exception>
+        public void SetConnectionString(string connectionString)
```

Comment:
it could be a setter on `ConnectionString` property - why does it have to be a separate method?

Reply:
1. In my opinion, it feels wrong that a property setter would have so many side effects.
I thought that an explicit method call would be more appropriate in this case. 

2. In the TelemetryConfigurationFactory, we use reflection to set all the properties on the TelemetryConfiguration object.
In our requirements, a config set ConnectionString should always overwrite a config set InstrumentationKey.
These properties are set in the order that they are discovered, so it's possible that the config IKey could overwrite the value from the config ConnString.
In my opinion, having an explicit method to call makes it more obvious that the ConnectionString would overwrite all other properties.
(See TelemetryConfigurationFactory line 482)

Reply:
Well, check out .NET. e.g. HTTP client https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/HttpClient.cs#L97 or https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/HttpClientHandler.Windows.cs#L207

the side effects of  setters are fine. 
Having two ways to set ikey IS confusing, but I don't see how separate method helps with it. 
It's even more confusing to have inconsistent ways to set ikey vs connection string.

Reply:
I still have the same concern :)